compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -351,6 +359,249 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
+//
+// clang-format on
+static std::optional<Value>
+matchCatNegateAndSlice(tensor::InsertSliceOp insertOp) {
+  /// First match against the desired op chain.
+  auto topHalf = insertOp.getSource().getDefiningOp<tensor::ExtractSliceOp>();
+  if (!topHalf) {
+    return std::nullopt;
+  }
+  auto dest = insertOp.getDest().getDefiningOp<tensor::InsertSliceOp>();
+  if (!dest) {
+    return std::nullopt;
+  }
+
+  /// TODO: This could be extended to other unary (or in general, elementwise
+  /// operations) if the need arises.
+  auto negate = dest.getSource().getDefiningOp<linalg::GenericOp>();
+  if (!negate || !isUnaryNegate(negate)) {
+    return std::nullopt;
+  }
+
+  auto bottomHalf =
+      negate.getDpsInputs()[0].getDefiningOp<tensor::ExtractSliceOp>();
+  if (!bottomHalf) {
+    return std::nullopt;
+  }
+  Value source = topHalf.getSource();
+  if (source != bottomHalf.getSource()) {
+    return std::nullopt;
+  }
+
+  /// Require that the overall operation isn't changing the tensor shape.
+  if (source.getType() != dest.getType()) {
+    return std::nullopt;
+  }
+
+  auto sourceType = source.getType().dyn_cast<RankedTensorType>();
+  if (!sourceType) {
+    return std::nullopt;
+  }
+
+  /// Here we are verifying that the slice sequence inserts and extracts in the
+  /// same order. In other words, the offsets, sizes, and strides for the nth
+  /// extract slice match the nth insert slice.
+  /// TODO: We could relax this condition if we see cases where the elementwise
+  /// unary operation happens on a different slices of the tensor, but for now
+  /// don't engineer for cases that don't exist.
+  auto isSame = [](OpFoldResult a, OpFoldResult b) { return a == b; };
+  if (!insertOp.isSameAs(bottomHalf, isSame) ||
+      !dest.isSameAs(topHalf, isSame)) {
+    return std::nullopt;
+  }
+
+  /// All slices need unit stride and the initial slice should have a zero
+  /// offset.
+  if (!insertOp.hasUnitStride() || !dest.hasUnitStride() ||
+      !dest.hasZeroOffset()) {
+    return std::nullopt;
+  }
+
+  /// Verify all slice sizes are the same.
+  for (auto [bottom, top] :
+       llvm::zip_equal(insertOp.getMixedSizes(), dest.getMixedSizes())) {
+    if (bottom != top) {
+      return std::nullopt;
+    }
+  }
+
+  /// Verify that the final InsertSlice has zero offsets for all except the
+  /// inner most dim.
+  SmallVector<OpFoldResult> insertOffsets = insertOp.getMixedOffsets();
+  for (int i = 0, e = insertOffsets.size() - 1; i < e; ++i) {
+    if (!isConstantIntValue(insertOffsets[i], 0)) {
+      return std::nullopt;
+    }
+  }
+
+  /// Finally verify that the inner most offset == the inner most slice size,
+  /// and because we verified all slice sizes are the same, this necessarily
+  /// is slicing exactly half of the inner most dimension.
+  if (insertOffsets.back() != insertOp.getMixedSizes().back()) {
+    return std::nullopt;
+  }
+
+  return source;
+}
+
+static void rewriteCatNegateAndSlice(RewriterBase &rewriter,
```

Comment:
Nit: To be consistent, make this return the replacement value, and do the replacement outside the function.